openedx/features/enterprise_support/enrollments/utils.py

```diff
@@ -38,32 +95,73 @@ def lms_enroll_user_in_course(
     Arguments:
      - username (str): User name
      - course_id (obj) : Course key obtained using CourseKey.from_string(course_id_input)
-     - mode (CourseMode): course mode
-     - enterprise_uuid (str): id to identify the enterprise to enroll under
-     - is_active (bool): Optional. A Boolean value that indicates whether the
+     - desired_mode (CourseMode): desired course mode
+     - is_active (bool): A Boolean value that indicates whether the
         enrollment is to be set to inactive (if False). Usually we want a True if enrolling anew.
+     - enterprise_uuid (str): Optional. id to identify the enterprise to enroll under
 
     Returns: A serializable dictionary of the new course enrollment. If it hits
      `CourseEnrollmentExistsError` then it logs the error and returns None.
     """
     user = _validate_enrollment_inputs(username, course_id)
+    current_enrollment = enrollment_api.get_enrollment(username, str(course_id))
 
     with transaction.atomic():
         try:
-            response = enrollment_api.add_enrollment(
+            if current_enrollment:
+                if (
+                    current_enrollment['mode'] != desired_mode
+                    or current_enrollment['is_active'] != is_active
+                ):
+                    response = enrollment_api.update_enrollment(
+                        username,
+                        str(course_id),
+                        mode=desired_mode,
+                        is_active=is_active,
+                        enrollment_attributes=None,
+                    )
+                    if not response or (
+                        response['mode'] != desired_mode or
+                        response['is_active'] != is_active
+                    ):
+                        log.exception(
+                            "An error occurred while updating the new course enrollment for user "
+                            "[%s] in course run [%s]",
+                            username,
+                            course_id,
+                        )
+                        raise EnrollmentNotUpdatableError(
+                            f"Unable to upgrade enrollment for user {username} "
+                            "in course {course_id} to {desired_mode} mode."
+                        )
+                return None
```

Comment:
[clarification] Just a sanity check about this `return None`. If an existing enrollment needs updating (i.e., passes the 2 conditionals above) and its update is successful, it appears this statement will short circuit the rest of the method such that no response is returned to the caller of the method.

I'm wondering if it might make sense  to return the updated enrollment response and update the docstring:

> Returns: A serializable dictionary of the new course enrollment. If it hits `CourseEnrollmentExistsError` then it logs the error and returns None.

... to account for also _updating_ an enrollment, not just returning a **new** course enrollment?

Reply:
That does make more sense! I changed the function so that it will return either the updated enrollment or the new course enrollment, unless it hits a CourseEnrollmentError or a CourseEnrollmentNotUpdatableError. I also set it to return None if the "upgrade" parameters are the same as the existing enrollment.